pyclam/criterion.py

```diff
@@ -337,3 +337,36 @@ def _select(root: Cluster, upper: float, lower: float, grace_depth: int) -> Set[
             inactive = {child for cluster in inactive for child in cluster.children}
 
         return selected
+
+
+class Labels(SelectionCriterion):
+    def __init__(self, labels: Dict[int, int]):
+        """ Uses a given dictionary of index -> label to select optimal clusters.
+
+        During a BFT of the tree, when a cluster has an overwhelming majority of points with the same label,
+        that cluster is selected.
+        If some data points are unlabelled, then when a cluster contains no labelled points, it too is selected.
+        """
+        self.labels: Dict[int, int] = labels
+        self.min_fraction: float = min(dict(Counter(labels.values())).values()) / len(labels)
+        self.threshold: float = 1 - (self.min_fraction / 10)
+
+    def __call__(self, root: Cluster) -> Set[Cluster]:
+        selected: Set[Cluster] = set()
+        frontier: Set[Cluster] = {root}
+        while frontier:
+            cluster = frontier.pop()
+            if cluster.children is None:
+                selected.add(cluster)
+            else:
+                labels: List[int] = [self.labels[p] for p in cluster.argpoints if p in self.labels]
+                if len(labels) > 0:
+                    max_fraction = max(dict(Counter(labels)).values()) / cluster.cardinality
+                    if max_fraction > self.threshold:
+                        selected.add(cluster)
+                    else:
+                        frontier.update(cluster.children)
+                else:
+                    # TODO: Is this really the best idea?
```

Comment:
Seems like the only good path. We can't expect that future children will suddenly have more labels. Unless you just take these types of situations down to singletons? What other ideas did you have?

Reply:
reverting to other selection criteria?